packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -440,11 +441,105 @@ class RenderParagraph extends RenderBox {
   @override
   void describeSemanticsConfiguration(SemanticsConfiguration config) {
     super.describeSemanticsConfiguration(config);
-    config
-      ..label = text.toPlainText()
-      ..textDirection = textDirection;
+    bool hasRecognizers = false;
+    text.visitTextSpan((TextSpan span) {
+      if (span.recognizer != null) {
+        hasRecognizers = true;
+        return false;
+      }
+      return true;
+    });
+    if (hasRecognizers) {
+      config.explicitChildNodes = true;
+      config.isSemanticBoundary = true;
+    } else {
+      config.label = text.toPlainText();
+      config.textDirection = textDirection;
+    }
   }
 
+  @override
+  void assembleSemanticsNode(SemanticsNode node, SemanticsConfiguration config, Iterable<SemanticsNode> children) {
+    final List<SemanticsNode> newChildren = <SemanticsNode>[];
+    StringBuffer buffer;
+    double order = 0.0;
+    int start = 0;
+    int end = 0;
+    // traverse all spans and collect them into adjacent compatible nodes.
+    text.visitTextSpan((TextSpan span) {
+      if (span.recognizer != null) {
+        if (buffer != null) {
+          final SemanticsNode child = new SemanticsNode();
+          final SemanticsConfiguration childConfig = new SemanticsConfiguration();
+          child.rect = _getRectForSpan(start, end);
+          childConfig
+            ..label = buffer.toString()
+            ..textDirection = textDirection
```

Comment:
this seems wrong. What if the text is bidi?

For example, consider the following text, where lowercase represents LTR text, uppercase represents RTL text, and square brackets represent links, `<RLE>` represents Unicode Character 'RIGHT-TO-LEFT EMBEDDING' (U+202B), and `<PDF>` represents Unicode Character 'POP DIRECTIONAL FORMATTING' (U+202C).

> `hello world <RLE> [BOY] HOW DO you DO [SIR] <PDF> good bye`

Your current code is going to split this into five boxes:

> `hello world <RLE>`
> `BOY`
> `HOW DO you DO`
> `SIR`
> `<PDF> good bye`

...which will result in the following (correct) rendering being painted:

> `hello world RIS OD you OD WOH YOB good bye`

...but the following rendering being described by the semantics:

> `hello world YOB OD WOH you OD RIS good bye`

...which is completely mangled.

Reply:
(I don't really know what the right fix for this is...)

Reply:
I will take the time to figure out the correct solution then

Reply:
We'll need a new semantics ordering algorithm specifically for text subsections - I have some thoughts that I'll put together into a short doc

Reply:
When collecting the text I can track the stack of directionality instructions and change whether I am pushing/inserting to a text buffer. However this will only work when I have explicit ltr or rtl embedding characters. So without exposing new engine APIs I believe I can do:

`hello world RIS OD uoy OD WOH YOB good bye`

Reply:
Update: I've made more progress towards this. I need to do more verification but I believe the strategy I have now will just work (tm)